File: 0022-generate-parentheses/0022-generate-parentheses.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void parenthesis(char **result,int *count,char *current,int n,int open,int close){
    if(strlen(current)==2*n){
        result[*count]=strdup(current);
        (*count)++;
        return;
    }
    if(open<n){
        char newStr[2*n+1];
        strcpy(newStr,current);
        strcat(newStr,"(");
        parenthesis(result,count,newStr,n,open+1,close);
    }
    if(close<open){
        char newStr[2*n+1];
        strcpy(newStr,current);
        strcat(newStr,")");
        parenthesis(result,count,newStr,n,open,close+1);
    }
}

char** generateParenthesis(int n, int* returnSize) {
    char **result=(char **)malloc(100000*sizeof(char *));
    *returnSize=0;
    char *start=(char *)malloc((2*n+1) * sizeof(char));
    start[0]='\0';  
    parenthesis(result, returnSize, start, n, 0, 0);
    free(start);  
    return result;
}

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        String cur = "";
        helper(n,ans,0,0,cur);
        return ans;
    }
    private void helper(int n , List<String> ans,int open ,int close,String cur){

        if(cur.length() == 2*n){
            
                ans.add(cur);
            
            return;
        }
        if(open < n){
            helper(n,ans,open+1,close,cur+"(");
        }
        if(close < open){
            helper(n,ans,open,close+1,cur+")");
        }

    }
}